src/app-admin.js

```diff
@@ -0,0 +1,491 @@
+    res
+      .status(200)
+      .send({ status: 'ok', data: { someDeletionsFailed: false } });
+  } else {
+    sendErrorResponse(res, 400, 'not-all-deleted', '');
+  }
+});
+
+app.get('/access', (req, res) => {
+  const fileId = req.query.fileId;
+  const session = validateUser(req, res);
+  if (!session) return;
+
+  const { id: fileIdInDb } = getFileById(fileId);
+  if (!fileIdInDb) {
+    sendErrorResponse(res, 400, 'invalid-file-id', 'File not found at server');
+    return false;
+  }
+
+  const accesses = getAccountDb().all(
+    `SELECT users.id as userId, user_name as userName, files.owner, display_name as displayName
+     FROM users
+     JOIN user_access ON user_access.user_id = users.id
+     JOIN files ON files.id = user_access.file_id
+     WHERE files.id = ? and (files.owner = ? OR 1 = ?)`,
+    [fileId, session.user_id, !isAdmin(session.user_id) ? 0 : 1],
+  );
+
+  res.json(accesses);
+});
+
+function checkFilePermission(fileId, userId, res) {
+  const { granted } = getAccountDb().first(
+    `SELECT 1 as granted
+      FROM files
+      WHERE files.id = ? and (files.owner = ?)`,
+    [fileId, userId],
+  ) || { granted: 0 };
+
+  if (granted === 0 && !isAdmin(userId)) {
+    sendErrorResponse(
+      res,
+      400,
+      'file-denied',
+      "You don't have permissions over this file",
+    );
+    return false;
+  }
+
+  const { id: fileIdInDb } = getFileById(fileId);
+  if (!fileIdInDb) {
+    sendErrorResponse(res, 400, 'invalid-file-id', 'File not found at server');
+    return false;
+  }
+
+  return true;
+}
+
+app.post('/access', (req, res) => {
+  const userAccess = req.body || {};
+  const session = validateUser(req, res);
+
+  if (!session) return;
+
+  if (!checkFilePermission(userAccess.fileId, session.user_id, res)) return;
+
+  if (!userAccess.userId) {
+    sendErrorResponse(res, 400, 'user-cant-be-empty', 'User cannot be empty');
+    return;
+  }
+
+  const { cnt } =
+    getAccountDb().first(
+      'SELECT count(*) AS cnt FROM user_access WHERE user_access.file_id = ? and user_access.user_id = ?',
+      [userAccess.fileId, userAccess.userId],
+    ) || {};
+
+  if (cnt > 0) {
+    sendErrorResponse(
+      res,
+      400,
+      'user-already-have-access',
+      'User already have access',
+    );
+    return;
+  }
+
+  getAccountDb().mutate(
+    'INSERT INTO user_access (user_id, file_id) VALUES (?, ?)',
+    [userAccess.userId, userAccess.fileId],
+  );
+
+  res.status(200).send({ status: 'ok', data: {} });
+});
+
+app.post('/access/delete-all', (req, res) => {
+  const fileId = req.query.fileId;
+  const session = validateUser(req, res);
+  if (!session) return;
+
+  if (!checkFilePermission(fileId, session.user_id, res)) return;
+
+  const ids = req.body.ids;
+  let totalDeleted = 0;
+  ids.forEach((item) => {
+    const accessDeleted = getAccountDb().mutate(
+      'DELETE FROM user_access WHERE user_id = ?',
+      [item],
+    ).changes;
+    totalDeleted += accessDeleted;
+  });
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏you can do a batch delete here

```
DELETE FROM user_access WHERE user_id IN ?
```